config/config.go

```diff
@@ -92,6 +113,71 @@ func Unmarshall(flowConfig *FlowConfig) error {
 	return nil
 }
 
+// Print prints current configuration keys and values.
+// Returns:
+// map[string]struct{}: map of keys to avoid printing if they were set by an config file.
+// This is required because we still have other config values not migrated to the config package. When a
+// config file is used currently only network-configs are set, we want to avoid printing the config file
+// value and also the flag value.
+func Print(info *zerolog.Event, flags *pflag.FlagSet) map[string]struct{} {
+	// only print config values if they were overridden with a config file
+	m := make(map[string]struct{})
+	if flags.Lookup(configFilePath).Changed {
+		for _, key := range conf.AllKeys() {
+			info.Str(key, fmt.Sprintf("%v", conf.Get(key)))
+			s := strings.Split(key, ".")
+			if len(s) == 2 {
+				m[s[1]] = struct{}{}
+			} else {
+				m[key] = struct{}{}
+			}
+		}
+	}
+
+	return m
+}
+
+// setAliases sets aliases for config sub packages. This should be done directly after pflags are bound to the configuration store.
+// Upon initialization the conf will be loaded with the default config values, those values are then used as the default values for
+// all the CLI flags, the CLI flags are then bound to the configuration store and at this point all aliases should be set if configuration
+// keys do not match the CLI flags 1:1. ie: networking-connection-pruning -> network-config.networking-connection-pruning. After aliases
+// are set the conf store will override values with any CLI flag values that are set as expected.
+func setAliases() {
+	err := network.SetAliases(conf)
+	if err != nil {
+		panic(fmt.Errorf("failed to set network aliases: %w", err))
+	}
+}
+
+// overrideConfigFile overrides the default config file by reading in the config file at the path set
+// by the --config-file and --config-file-name flags in our viper config store.
+//
+// Returns:
+//
+//	error: if there is any error encountered while reading new config file, all errors are considered irrecoverable.
+//	bool: true if the config was overridden by the new config file, false otherwise or if an error is encountered reading the new config file.
+func overrideConfigFile(flags *pflag.FlagSet) (error, bool) {
+	configFileFlag := flags.Lookup(configFilePath)
+	if configFileFlag.Changed {
+		p := configFileFlag.Value.String()
+		dirPath, fileName := splitConfigPath(p)
+		conf.AddConfigPath(dirPath)
+		conf.SetConfigName(fileName)
+		err := conf.ReadInConfig()
+		if err != nil {
+			return fmt.Errorf("failed to read config file %s: %w", p, err), false
+		}
+		return nil, true
+	}
+	return nil, false
+}
+
+// getConfigNameFromPath returns the directory and name of the config file from the provided path string.
+func splitConfigPath(path string) (string, string) {
```

Comment:
In this function, you are trying to split the path into the directory and the base name of the config file without the extension. This is fine, but it’s good to note that if someone names a config file with multiple dots (e.g., `my.config.yaml`), this function would consider `my` as the name which might not be intended. I suggest the following implementation that uses a regular expression pattern and panics if the input path does not match the pattern: 

```go
// splitConfigPath returns the directory and base name (without extension) of the config file from the provided path string.
// If the file name does not match the expected pattern, the function panics.
//
// The expected pattern for file names is that they must consist of alphanumeric characters, hyphens, or underscores,
// followed by a single dot and then the extension.
//
// Legitimate Inputs:
//     - /path/to/my_config.yaml
//     - /path/to/my-config123.yaml
//     - my-config.yaml (when in the current directory)
//
// Illegitimate Inputs:
//     - /path/to/my.config.yaml (contains multiple dots)
//     - /path/to/my config.yaml (contains spaces)
//     - /path/to/.config.yaml (does not have a file name before the dot)
//
// Args:
//     - path: The file path string to be split into directory and base name.
//
// Returns:
//     - The directory and base name without extension.
//
// Panics:
//     - If the file name does not match the expected pattern.
func splitConfigPath(path string) (string, string) {
	// Regex to match filenames like 'my_config.yaml' or 'my-config.yaml' but not 'my.config.yaml'
	validFileNamePattern := regexp.MustCompile(`^[a-zA-Z0-9_-]+\.[a-zA-Z0-9]+$`)

	dir, name := filepath.Split(path)

	// Panic if the file name does not match the expected pattern
	if !validFileNamePattern.MatchString(name) {
		panic(fmt.Errorf("Invalid config file name '%s'. Expected pattern: alphanumeric, hyphens, or underscores followed by a single dot and extension", name))
	}

	// Extracting the base name without extension
	baseName := strings.Split(name, ".")[0]
	return dir, baseName
}

```

Reply:
https://github.com/onflow/flow-go/pull/4340/commits/99d7a7884da874e9bb274a5a47de30e604d11a15